File: scripts/dal_dump_app_config.py
#!/usr/bin/env python

class TableRow :
    def __init__(self, fill, separator, s1, s2, s3, s4, s5, s6):
        self.fill = fill
        self.separator = separator
        self.items = []
        if s1 is not None :
            self.items.append(s1)
            self.items.append(s2)
            self.items.append(s3)
            self.items.append(s4)
            self.items.append(s5)         
            self.items.append(s6)  
        else :
            self.items.append("=")
            self.items.append("=") 
            self.items.append("=") 
            self.items.append("=") 
            self.items.append("=") 
            self.items.append("=") 
                         
    @classmethod
    def fromhost(cls, fill, separator, host ):
        return cls(fill, separator, "", "", host.id + "@" + host.class_name , "", "", "")
    @classmethod
    def fromobjects(cls, fill, separator, num, app, host, seg, seg_id, app_id):
        return cls(fill, separator, num, app.id + "@" + app.class_name, host.id + "@" + host.class_name, seg.id + "@" + seg.class_name, seg_id, app_id )
    @classmethod
    def fromchar(cls, fill, separator):
        return cls(fill, separator, None,None,None,None,None,None)

  
if __name__ == '__main__':

    import argparse
    import sys    
     
    import oksdbinterfaces

    import dal

        
    parser = argparse.ArgumentParser(description='''This program prints out applications and their configuration parameters using Partition::get_all_applications() algorithm.
        Usage: dal_dump_app_config.py [-d database-name] -p name-of-partition [-t [types ...]] [-c [ids ...]] [-s [ids ...]]''')
    parser.add_argument("-d", "--database_name", 
                        help="name of the database (ignore TDAQ_DB variable)")
    parser.add_argument("-p", "--partition_name", required = True,
                        help="name of the partition object")
    parser.add_argument("-t", "--application_types",   nargs = '*',
                        help="filter out all applications except given classes (and their subclasses)")
    parser.add_argument("-c", "--hosts",   nargs = '*',
                        help="filter out all applications except those which run on given hosts")
    parser.add_argument("-s", "--segments", nargs = '*',
                        help="filter out all applications except those which belong to given segments")  
    parser.add_argument("-b", "--show_backup_hosts", action='count',
                        help="print backup hosts")                       

    args = parser.parse_args()

      # Open test database
    db = oksdbinterfaces.Configuration(args.database_name)

    # get Partition object
    partition = db.get_dal('Partition', args.partition_name)

    app_config_list = partition.get_all_applications(db, args.application_types , args.segments, args.hosts)

    print ('Got ' + str(len(app_config_list)) + ' applications:')

    rows = []
    rows.append(TableRow.fromchar('=', '='))
    rows.append(TableRow(' ', '|', "num","Application Object","Host","Segment","Segment unique id","Application unique id"))    
    rows.append(TableRow.fromchar('=', '='))    

    count = 1

    for i in app_config_list :
#         init = i.get_initialization_depends_from(app_config_list)
#         print "initialisation of",i.get_app_id(),"depends from",len(init)
#         shut = i.get_shutdown_depends_from(app_config_list)
#         for x in init:
#             print '   ',x.get_app_id(),'on',x.get_host().id
#         print "shutdown of",i.get_app_id(),"depends from",len(shut)
#         for x in shut:
#             print '   ',x.get_app_id(),'on',x.get_host().id
        rows.append(TableRow.fromobjects(' ', '|', str(count), i.get_base_app(), i.get_host(), i.get_base_seg(), i.get_seg_id(), i.get_app_id()))         
        count+=1
        if args.show_backup_hosts is not None:
            backup_hosts = i.get_backup_hosts()
            for j in backup_hosts :
                rows.append(TableRow.fromhost(' ', '|', j))  

    rows.append(TableRow.fromchar('=', '=')) 

    cw = [1,1,1,1,1,1]

    for row in rows:
        #print row.items
        for idx in range(0,6) : 
            length = len(str(row.items[idx]))
            if length > cw[idx]:
                cw[idx] = length
        #print cw

    align_left = [False, True, True, True, True, True]

    for row in rows :
        line =[]
        for idx in range(0,6):
            line.append(row.separator)
            line.append(row.fill)
            if align_left[idx]:
                line.append(row.items[idx])
                line.append(row.fill*(cw[idx]-len(row.items[idx]))) 
                line.append(row.fill)
            else :
                line.append(row.fill*(cw[idx]-len(row.items[idx])))        
                line.append(row.items[idx])        
                line.append(row.fill)

        line.append(row.separator)     
        line = ''.join(str(e) for e in line)  
        print(line)
                               

File: apps/dal_dump_app_config.cxx
#include <sstream>

#include <boost/program_options.hpp>

#include "oksdbinterfaces/Configuration.hpp"

#include "dal/util.hpp"

#include "dal/Partition.hpp"
#include "dal/BaseApplication.hpp"
#include "dal/Computer.hpp"
#include "dal/Segment.hpp"


////////////////////////////////////////////////////////////////////////////////////////////////////////////////

  // The type TableRow describes table's row with 6 items, separator and fill chars.
  // It is used to print separator line, title row and application parameters

struct TableRow
{
  char m_fill;
  char m_separator;
  std::string m_items[6];
  TableRow(const char fill, const char separator) :
      m_fill(fill), m_separator(separator)
  {
    ;
  }

  TableRow(const char fill, const char separator, unsigned short num, const dunedaq::dal::BaseApplication * app, const dunedaq::dal::Computer * host, const dunedaq::dal::Segment * seg, const std::string& seg_id, const std::string& app_id) :
      m_fill(fill), m_separator(separator)
  {
    std::ostringstream num_s;
    num_s << num;
    m_items[0] = num_s.str();
    m_items[1] = app->UID() + '@' + app->class_name();
    if (!DalObject::is_null(host))
      m_items[2] = host->UID() + '@' + host->class_name();
    else
      m_items[2] = "(null)";
    m_items[3] = seg->UID() + '@' + seg->class_name();
    m_items[4] = seg_id;
    m_items[5] = app_id;
  }

  TableRow(const char fill, const char separator, const dunedaq::dal::Computer& host) :
      m_fill(fill), m_separator(separator)
  {
    m_items[0] = m_items[1] = m_items[3] = m_items[4] = m_items[5] = "";
    m_items[2] = host.UID() + '@' + host.class_name();
  }

  TableRow(const char fill, const char separator, const std::string& s1, const std::string& s2, const std::string& s3, const std::string& s4, const std::string& s5, const std::string& s6) :
      m_fill(fill), m_separator(separator)
  {
    m_items[0] = s1;
    m_items[1] = s2;
    m_items[2] = s3;
    m_items[3] = s4;
    m_items[4] = s5;
    m_items[5] = s6;
  }
};

////////////////////////////////////////////////////////////////////////////////////////////////////////////////

int
main(int argc, char *argv[])
{

  std::string data;
  std::string partition_name;
  std::set<std::string> app_types;
  std::vector<std::string> hosts_list;
  std::set<const dunedaq::dal::Computer *> hosts;
  std::set<std::string> segments;
  bool backup = false;

  boost::program_options::options_description desc("This program prints out applications and their configuration parameters using Partition::get_all_applications() algorithm");

  try
    {
      std::vector<std::string> app_types_list;
      std::vector<std::string> segments_list;

      desc.add_options()
        ("data,d", boost::program_options::value<std::string>(&data), "name of the database")
        ("partition-name,p", boost::program_options::value<std::string>(&partition_name)->required(), "name of the partition object")
        ("application-types,t", boost::program_options::value<std::vector<std::string> >(&app_types_list)->multitoken(), "filter out all applications except given classes (and their subclasses)")
        ("hosts,c", boost::program_options::value<std::vector<std::string> >(&hosts_list)->multitoken(), "filter out all applications except those which run on given hosts")
        ("segments,s", boost::program_options::value<std::vector<std::string> >(&segments_list)->multitoken(), "filter out all applications except those which belong to given segments")
        ("show-backup-hosts,b","print backup hosts")
        ("help,h", "Print help message");

      boost::program_options::variables_map vm;
      boost::program_options::store(boost::program_options::parse_command_line(argc, argv, desc), vm);

      if (vm.count("help"))
        {
          std::cout << desc << std::endl;
          return EXIT_SUCCESS;
        }

      boost::program_options::notify(vm);

      if (vm.count("show-backup-hosts"))
        {
          backup = true;
        }

      for (const auto& x : app_types_list)
        app_types.insert(x);

      for (const auto& x : segments_list)
        segments.insert(x);
    }
  catch (std::exception& ex)
    {
      std::cerr << "Command line parsing errors occurred:\n" << ex.what() << std::endl;
      return EXIT_FAILURE;
    }

  try
    {
      Configuration db(data);

      const dunedaq::dal::Partition * partition = dunedaq::dal::get_partition(db, partition_name);

      if (!partition)
        return EXIT_FAILURE;

      db.register_converter(new dunedaq::dal::SubstituteVariables(*partition));

      std::set<const dunedaq::dal::Computer *> hosts;

      for (const auto& x : hosts_list)
        {
          if (const dunedaq::dal::Computer * host = db.get<dunedaq::dal::Computer>(x))
            {
              hosts.insert(host);
            }
          else
            {
              std::cerr << "ERROR: cannot find computer \'" << x << "\'\n";
              return (EXIT_FAILURE);
            }
        }

      std::vector<const dunedaq::dal::BaseApplication *> apps = partition->get_all_applications((app_types.empty() ? nullptr : &app_types), (segments.empty() ? nullptr : &segments), (hosts.empty() ? nullptr : &hosts));

      std::cout << "Got " << apps.size() << " applications:\n";

      std::vector<TableRow> rows;
      rows.reserve(apps.size()+4);

      // fill the table
        {
          rows.push_back(TableRow('=', '='));
          rows.push_back(TableRow(' ', '|', "num","Application Object","Host","Segment","Segment unique id","Application unique id"));
          rows.push_back(TableRow('=', '='));

          unsigned long count = 1;
          for (const auto& i : apps)
            {
              rows.push_back(TableRow(' ', '|', count++, i->get_base_app(), i->get_host(), i->get_segment()->get_base_segment(), i->get_segment()->UID(), i->UID()));

              if (backup)
                {
                  for (const auto& x : i->get_backup_hosts())
                    {
                      rows.push_back(TableRow(' ', '|', *x));
                    }
                }
            }

          rows.push_back(TableRow('=', '='));
        }

      // detect columns widths
      unsigned short cw[] = {1,1,1,1,1,1};

      for (const auto& i: rows)
        {
          for (unsigned short j = 0; j < sizeof(cw)/sizeof(unsigned short); ++j)
            {
              unsigned short len = i.m_items[j].size();

              if(len > cw[j])
                cw[j]=len;
            }
        }


        // print table
      bool align_left[] = {false, true, true, true, true, true};

      for (const auto& i: rows)
        {
          std::cout.fill(i.m_fill);
          for (unsigned short j = 0; j < sizeof(cw)/sizeof(unsigned short); ++j)
            {
              std::cout << i.m_separator << i.m_fill;
              std::cout.setf((align_left[j] ? std::ios::left : std::ios::right), std::ios::adjustfield);
              std::cout.width(cw[j]);
              std::cout << i.m_items[j] << i.m_fill;
            }
          std::cout << i.m_separator << std::endl;
        }
    }
  catch (ers::Issue & ex)
    {
      std::cerr << "Caught " << ex << std::endl;
      return (EXIT_FAILURE);
    }

  return EXIT_SUCCESS;
}
